Feat: multiple parameter types

kilianmh opened the pull request:
Support parameters with multiple types.
For each type a separate specialized methods is defined.
When there are multiple parameters with more than one type,
all possible combination methods are defined.

Also: add type declarations, docstrings to changed functions.

It would be probably better to include unit tests, but I was not sure how you would like that. Probably better in a separate PR.

Closes #3

actions-user committed:
Update docs

kilianmh committed:
Feat: multiple parameter types

Support parameters with multiple types.
For each type a separate specialized methods is defined.
When there are multiple parameters with more than one type,
all possible combination methods are defined.

Also: add type declarations, docstrings to changed functions.

svetlyak40wt commented on client/core.lisp:
```diff
@@ -57,16 +57,25 @@
+        (if (stringp type)
+            (%schema-to-type type)
+            (mapc #'%schema-to-type type)))
+      type-list))
```
It is better to apply `nreverse` to `type-list`, because otherwise types will in opposite direction.

svetlyak40wt commented on client/core.lisp:
```diff
@@ -85,28 +94,68 @@
+  (declaim (ftype (function (cons) cons) expand-list))
+  (defun expand-list (lambda-list)
+    "Expand required parameters to all possible combinations of types.
+     If required parameters have one type, the result is same as input."
```
This function is a little bit tricky. Could you please add a unit-test checking if it behaves correctly on inputs with multiple types?

kilianmh replied:
Indeed, this was the most tricky part in this PR. Now some unit tests are there for `expand-list`.

svetlyak40wt commented on client/core.lisp:
```diff
@@ -267,9 +316,30 @@
+    (flet ((%generate-defmethod (&optional lambda-list-element)
+             `(defmethod ,name ,(cons `(client ,class-name) lambda-list-element)
+                          (let* ((args ,arguments-collector))
+                            (labels ((retrieve-data (args)
```
Here `args` argument of `retrieve-data` shadows outer `args` binding in the `let*`. It is better to choose another name.

Also, it is better to use `gensym` for local variables of the `defmethod`.

svetlyak40wt commented on client/core.lisp:
```diff
@@ -289,22 +359,16 @@
-                                    (let ((raw-response (rpc-call client ,original-name args)))
-                                      ,result-transformation)))
-                           (retrieve-data args)))))))
+                    (generate-defmethod name class-name lambda-list arguments-collector
```
Probably it is better to rename `generate-defmethod` to `generate-defmethods` because now it returns a list.

kilianmh committed:
Fix: gensym in defmethods

Also:
- reverse type-list in schema-to-type so that order is as in spec
- add unit-tests for expand-list
- rename defmethod to defmethods

kilianmh commented:
Now the requested changes are done. Anything else that should be improved? I can also squash the changes if you want that.

svetlyak40wt committed:
Add a simple regression test.

svetlyak40wt commented:
I'm trying this pull on a simple spec and client is not able to generate code. It is failing with "Type "null" is not supported yet." error pointing to this line: https://github.com/40ants/openrpc/pull/11/files#diff-4a997ff74e9cd54906e0eed1815f4f93c2ad21f5332819b9c831a62e650aa1f5R73

After I've added the code: `((string-equal type "null") (push 'null type-list))` macro started to generate another error:

```
The value
  ("string" "null")
is not of type
  (OR STRING SYMBOL CHARACTER)
when binding SB-IMPL::STRING1
```

I've added this commit https://github.com/40ants/openrpc/pull/11/commits/8b2cb719559b43606d85f50157913190f050ebe2 to the pull request. Could you please, try to make it work?

kilianmh committed:
Fix: type error

Avoid string-equal on list with types.

kilianmh commented:
- The type error should now be fixed
- The tests are still not succeeding and it seems like `class-definitions` is always `nil`
```lisp
 Expect OPENRPC-TESTS/CLIENT/REGRESSION::CLASS-DEFINITIONS to be true.
     OPENRPC-TESTS/CLIENT/REGRESSION::CLASS-DEFINITIONS
```
Not sure how to how to proceed from here ... maybe remove `class-definitions` from `%generate-client` or from the unit-test?

svetlyak40wt commented:
Oh, this is because there is no any class definitions in the specification. I'll update the test to check for generated code instead of the fact it was generated.

kilianmh committed:
Fix: Simplify expand-list

The function expand-list now takes input as list such as
'((beta integer) (delta number string))

The test cases have been adjusted accordingly.

This should reduce consing slightly and simplify reasoning about the
working of expand-list.

svetlyak40wt committed:
Regression test was finished and now checks generated code against data in the files.

svetlyak40wt committed:
Made it easier to fix regression.

svetlyak40wt committed:
Update QL file and add some trace.

svetlyak40wt committed:
Trace where we installed bordeaux threads.

svetlyak40wt committed:
Trying with latest ultralisp.

svetlyak40wt committed:
Installing clack from quicklisp.

svetlyak40wt committed:
Updated CI file to use clack from quicklisp.

svetlyak40wt committed:
Move to newest Ultralisp where clack was fixed.

svetlyak40wt committed:
Move to newer Ultralisp where no dependency from uiop for 40ants-asdf-system.

kilianmh commented:
Damn, github actions is [dirty](https://news.ycombinator.com/item?id=37612420) ...

svetlyak40wt committed:
Removed quicklisp from CI matrix.

svetlyak40wt commented:
Well, tests are green now and this PR can be merged. What do you think?

svetlyak40wt commented:
Oh, one more thing! We need to write an update for the ChangeLog: https://github.com/40ants/openrpc/blob/master/docs/changelog.lisp

kilianmh commented:
Sure I can update the changelog. Should I add it to version `0.9.0` as `Additions` or create a new version (if yes, then 0.9.1 or 0.10.0)?

svetlyak40wt commented:
Let's make it 0.9.1, because this modification is more a bugfix than a feature. And it should break users.

kilianmh committed:
Docs: Changelog version 0.9.1

Add info to changelog about multiple parameter types fix.

kilianmh commented:
Changelog information is added now. 
Only caveat of merging this `PR` is that testing does not work on other implementations than `sbcl` but that seems (for me at least) more related to the test system than to the changes to the code.

svetlyak40wt committed:
Trying to run tests for some other lisp implementations.

svetlyak40wt commented:
> Only caveat of merging this PR is that testing does not work on other implementations than sbcl

I've updated CI pipeline to run tests for CCL, CLISP and ECL. But some of them fail
We can add more implementation if you think this could be useful.

svetlyak40wt committed:
Trying to mark some matrix combinations as experimental to prevent them from failing tests in PR.

As described at: https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#handling-failures

svetlyak40wt committed:
Fixed wrong error.

svetlyak40wt committed:
Added missed comma.

svetlyak40wt committed:
Wrapped with doublequotes booleans.

svetlyak40wt committed:
One more attempt.

svetlyak40wt committed:
Wrapped again.

svetlyak40wt committed:
Added os to ccl-bin include.

svetlyak40wt committed:
Exclude some matrix combinations.

svetlyak40wt committed:
Ignore ccl on ubuntu too.

svetlyak40wt committed:
Removed experimental tag from ECL on OSX.

kilianmh commented:
`ccl` is [failing here](https://github.com/40ants/openrpc/actions/runs/6292082050/job/17081123513?pr=11#step:8:206) not because the generated code is different but it is printed in a different way compared to `sbcl`. 

Particularly it prints everything in one line. This is not nice when inspecting generated code e.g. with `macroexpand`. Maybe its a good idea to format the generated code somehow (e.g. pretty print or some special formating library ?), but not sure if this should be part of this `PR`

kilianmh committed:
Fix: Pretty print in compare function

Now the generated lisp-code and the lisp-file are both pretty printed.

kilianmh commented:
Now the the tests work on CCL also. Are 2 implementations enough (for now) or we try to get more running?

svetlyak40wt commented:
@kilianmh you've made nice solution to data comparison for different lisp implementations.

I tried to use Incless for portable printing, but seems it does not support pretty printing yet: https://github.com/s-expressionists/Incless/issues/9

I'll see how to exclude clisp on macos for now as seems it fails because does not support threading.

svetlyak40wt committed:
Exclude clisp on osx.

svetlyak40wt committed:
Disabled ECL on OSX

svetlyak40wt commented:
@kilianmh thank you very much for your work. I'll add you to the "Contributors" section of the docs if you don't mind.

By the way, I have a list of other issues and improvements for this library. They are not filed as issues at the GitHub, but if you would like to continue contributing to it, I'll add them to the GitHub.

kilianmh commented:
> I'll add you to the "Contributors" section of the docs if you don't mind.

Sure you can add me to the contributors section :)

> if you would like to continue contributing to it, I'll add them

Yeah that would be good. I also have some ideas and will open issues for that if its not on your list.
